sqllogictest/src/runner.rs

```diff
@@ -303,12 +306,78 @@ impl<D: AsyncDB> Runner<D> {
 
     /// Run a sqllogictest script.
     pub fn run_script(&mut self, script: &str) -> Result<(), TestError> {
-        future::block_on(self.run_script_async(script))
+        block_on(self.run_script_async(script))
     }
 
     /// Run a sqllogictest file.
     pub fn run_file(&mut self, filename: impl AsRef<Path>) -> Result<(), TestError> {
-        future::block_on(self.run_file_async(filename))
+        block_on(self.run_file_async(filename))
+    }
+
+    /// aceept the tasks, spawn jobs task to run slt test. the tasks are (AsyncDB, slt filename)
+    /// pairs.
+    pub async fn run_parallel_async<F, Fut>(
+        &mut self,
+        glob: &str,
+        hosts: Vec<String>,
+        conn_builder: F,
+        jobs: usize,
+    ) -> Result<(), TestError>
+    where
+        F: Fn(String, String) -> Fut,
+        Fut: Future<Output = D>,
+    {
+        let files = glob::glob(glob).expect("failed to read glob pattern");
+        let mut tasks = vec![];
+        let conn_builder = Arc::new(conn_builder);
+
+        for (idx, file) in files.enumerate() {
+            // for every slt file, we create a database against table conflict
+            let file = file.unwrap();
+            let db_name = file
+                .file_name()
+                .expect("not a valid filename")
+                .to_str()
+                .expect("not a UTF-8 filename");
+            let db_name = db_name
+                .replace(' ', "_")
+                .replace('.', "_")
+                .replace('-', "_");
+
+            self.db
+                .run(&format!("CREATE DATABASE {};", db_name))
+                .await
+                .expect("create db failed");
+            let conn_builder = conn_builder.clone();
+            let target = hosts[idx % hosts.len()].clone();
+            tasks.push(async move {
+                let db = conn_builder(target, db_name).await;
+                let mut tester = Runner::new(db);
+                let filename = file.to_string_lossy().to_string();
+                tester.run_file_async(filename).await
+            })
+        }
+
+        let mut tasks = stream::iter(tasks).buffer_unordered(jobs);
+        while let Some(result) = tasks.next().await {
+            result?;
```

Comment:
The error returned here might be confusing to developers. As we are run-in a lot of test cases in parallel, we should let developer know which test case failed exactly. Better to have a new error type with which test case failed.